lib/web/files.go

```diff
@@ -0,0 +1,234 @@
+	Login string
+	// Namespace is node namespace.
+	Namespace string
+	// Cluster is the name of the remote cluster to connect to.
+	Cluster string
+	// Cluster is the name of the remote cluster to connect to.
+	RemoteFilePath string
+}
+
+// changePassword updates users password based on the old password
+func (h *Handler) uploadFile(w http.ResponseWriter, r *http.Request, p httprouter.Params, ctx *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
+	req := fileTransferRequest{
+		Login:          p.ByName("login"),
+		Namespace:      p.ByName("namespace"),
+		Server:         p.ByName("node"),
+		RemoteFilePath: p.ByName("filepath"),
+	}
+
+	clt, err := ctx.GetUserClient(site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ft := fileTransfer{
+		ctx:           ctx,
+		authClient:    clt,
+		proxyHostPort: h.ProxyHostPort(),
+	}
+
+	err = ft.upload(req, r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return ok(), nil
+}
+
+func (h *Handler) downloadFile(w http.ResponseWriter, r *http.Request, p httprouter.Params, ctx *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
+	req := fileTransferRequest{
+		Login:          p.ByName("login"),
+		Namespace:      p.ByName("namespace"),
+		Server:         p.ByName("node"),
+		RemoteFilePath: p.ByName("filepath"),
+	}
+
+	clt, err := ctx.GetUserClient(site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ft := fileTransfer{
+		ctx:           ctx,
+		authClient:    clt,
+		proxyHostPort: h.ProxyHostPort(),
+	}
+
+	err = ft.download(req, w)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return ok(), nil
+}
+
+type fileTransfer struct {
+	// ctx is a web session context for the currently logged in user.
+	ctx           *SessionContext
+	authClient    auth.ClientI
+	proxyHostPort string
+}
+
+func (f *fileTransfer) download(req fileTransferRequest, w http.ResponseWriter) error {
+	remoteDest, err := f.resolveRemoteDest(req.RemoteFilePath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tc, err := f.createClient(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	cmd, err := scp.CreateHTTPDownloadCommand(remoteDest, w, tc.Stdout)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = tc.RunSCPCommand(context.TODO(), cmd)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (f *fileTransfer) upload(req fileTransferRequest, httpReq *http.Request) error {
+	remoteDest, err := f.resolveRemoteDest(req.RemoteFilePath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tc, err := f.createClient(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	cmd, err := scp.CreateHTTPUploadCommand(remoteDest, httpReq, tc.Stdout)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = tc.RunSCPCommand(context.TODO(), cmd)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (f *fileTransfer) createClient(req fileTransferRequest) (*client.TeleportClient, error) {
```

Comment:
wait, why not use the client from the context? there is a  function there that does pretty much the same

Reply:
This behavior is similar to our web handlers for handling ssh connections. This method returns TeleportClient which then we use to connect to a node to run SCP command.

Reply:
k, can you then try to factor it into the common utility function to reduce boilerplate?

Reply:
i looked at 2 places where we create this config and moved initialization of common parameters to a separate method `makeTeleportClientConfig`